### PR TITLE
Fixes #19937 - Better job cancel handling

### PR DESCRIPTION
**Pull request by adamruzicka**
requires https://github.com/theforeman/foreman-tasks/pull/267

### CODE REVIEW

File: app/helpers/remote_execution_helper.rb [hunk 1]
```diff
@@ -122,6 +122,11 @@ def job_invocation_task_buttons(task)
                          :title => _('Try to cancel the job'),
                          :disabled => !task.cancellable?,
                          :method => :post)
+      buttons << link_to(_('Abort Job'), abort_foreman_tasks_task_path(task),
+                         :class => 'btn btn-danger',
+                         :title => _('Try to abort the job'),
```
Comment (iNecas) [inline app/helpers/remote_execution_helper.rb]: The title is not very descriptive, what about `Abort the the job: trying to cancel without waiting for the results from the remote hosts)` or something like that?

**Comment by iNecas** [comment]
I was still able to get into non-cancellable state (see screenshots).

What I did:

1. spawn a remote execution job
2. restart the rails server as well as the smart proxy while the job is running

non-cancellable-task-1
non-cancellable-task-2

Although the job is marked as failed, the host jobs are marked as running.

What was the reproducer you were running it against?

**Comment by adamruzicka** [comment]
I went with create a job, wait for the tasks to show up on the proxy, stop the proxy, try to cancel the job. Guess I got sidetracked with events not being propagated to sub plans.

If we want to tackle the rails restart too, we need to revisit resuming. Current "if everything failed in plan start over else fail and dont care about subplans any more" behavior obviously doesn't work here.

**Comment by iNecas** [comment]
What if we just started polling the sub-tasks in that case?

**Comment by adamruzicka** [comment]
The issue you described is now handled in https://github.com/theforeman/foreman-tasks/pull/267

**Comment by adamruzicka** [comment]
The issue you described is now handled in https://github.com/theforeman/foreman-tasks/pull/267

**Comment by iNecas** [comment]
Thanks, I can confirm it works for me after the change. The last thing I would suggest is improving the error messaging on the abort, currently I get something like this. The user probably doesn't care about the exception class, neither it's true it failed to initialize.

job-abort

**Comment by adamruzicka** [comment]
screenshot_20170801_123031 is this good enough or should I tweak the message a bit more?

**Comment by iNecas** [comment]
That would work for me

File: app/lib/actions/remote_execution/run_host_job.rb [hunk 2]
```diff
@@ -89,20 +90,28 @@ def continuous_output_providers
...
       def fill_continuous_output(continuous_output)
-        fill_planning_errors_to_continuous_output(continuous_output) unless exit_status
+        final_timestamp = (continuous_output.last_timestamp || task.ended_at).to_f + 1
+
+        if task.state == 'stopped' && task.result == 'cancelled'
+          continuous_output.add_output(_('Job cancelled by user'), 'debug', final_timestamp)
+        else
+          fill_planning_errors_to_continuous_output(continuous_output) unless exit_status
+        end
+
```
Comment (iNecas) [inline app/lib/actions/remote_execution/run_host_job.rb]: Please add this code after the line 103-105, to get the error messages about task cancelling to the end of the output:

```ruby
        delegated_output.fetch('result', []).each do |raw_output|
          continuous_output.add_raw_output(raw_output)
        end

        final_timestamp = (continuous_output.last_timestamp || task.ended_at).to_f + 1

        if task.state == 'stopped' && task.result == 'cancelled'
          continuous_output.add_output(_('Job cancelled by user'), 'debug', final_timestamp)
        else
          fill_planning_errors_to_continuous_output(continuous_output) unless exit_status
        end
```

**Comment by iNecas** [comment]
Thanks @adamruzicka

**Pull request merged**